src/contrib/mmap_cache.rst

```diff
@@ -0,0 +1,506 @@
+the GIDs of the groups the user is a member of. After the GID list the two names
+can be found as 0-terminated strings, first the string ``unique_name`` is
+pointing to and then the string for ``name``. The length of both strings
+including the terminating 0 byte is stored in ``strs_len`` and ``data_len``
+stores the length of all data, GIDs and strings, so it is ``data_len = strs_len
+* num_groups *sizeof(uint32_t)``.
+
+The second name attribute was added to mitigate a general issue the memory
+mapped cache currently has with lookup by names for initgr requests. A user can
+only have a single UID and a group can only have a single GID. A different UID
+would by definition automatically mean a different user from the point of view
+of the Linux kernel. User and group names are basically labels to the UID or
+GID, respectively, and multiple names can be assigned to a single UID or GID.
+Traditionally there are ``struct passwd`` and ``struct group`` to map the UID or
+GID with a single name and the ``getpwnam``, ``getpwuid``, ``getgrnam`` and
+``getgrgid`` calls are used to find the ID for a name and vice versa. But there
+is not restriction that the names used as first argument to ``getpwnam`` or
+``getgrnam`` have to be the same as the ones returned as ``pw_name`` or
+``gr_name`` in the related structs. ``pw_name`` and ``gr_name`` returned in
+``struct passwd`` and ``struct group`` respectively can be considered a canonical
+names. The names used as first argument for ``getpwnam`` and ``getgrnam``, as
+long as they differ from the canonical name, can be considered as alias
+names.
+
+Coming back to the memory mapped cache. The user and group data in the memory
+cache only contain a single name, the canonical name. This means a user or group
+entry can be only found in the memory mapped cache if the canonical name is used
+to lookup the entry. There is not an issue with many Unix/Linux based use case
+in general, but if e.g. the users are managed in Active Directory there might be
+different expectations about the name format, see e.g. `MSDN: User Name
+Formats`_. Besides the short *logon name* the fully-qualified *user principal
+name (UPN)* or the old NT style *down-level logon name* can be used. What makes
+it even worse is that names in AD are treated case-insensitive.
+
+.. _MSDN\: User Name Formats: https://msdn.microsoft.com/de-de/library/windows/desktop/aa380525(v=vs.85).aspx
+
+To allow lookups with different input names (aliases) one might be tempted to
+just replace the canonical name in the memory cache record with the input name.
+But this would fail if e.g. the related user is deleted on the server and has to
+be deleted in the memory mapped cache as well. Now all entries must be checked
+if they are somehow related to the deleted entry. To avoid this the canonical
+name is added as well and its hash is written to the otherwise unused ``hash2``
+element of ``struct sss_mc_rec``. Now can use different alias name to look up an
```

Comment:
Fix wording of `Now can use different alias name`

Reply:
I don't see  a difference.

Reply:
`Now different alias names can be used to look up an...`

Reply:
Done.